### PR TITLE
Add output_format_dependency()

### PR DESCRIPTION
**Pull request by atusy**
This PR is an attempt to solve #2347, but in a more generalized way.

In the issue #2347, I suggested `pandoc_dependency()`, however, this function needs to follow the updates of Pandoc's API changes.

Instead, this PR implements `output_format_dependency()`, which allows defining some of the arguments in `output_format` from within a Rmd file.
Supported arguments are `pre_processor`, `post_processor`, `file_scope`, `on_exit`.
The others are not because dependency can be solved after `post_knit`, which finalizes `knit_meta`.

Fortunately, `pre_processor` can return extra arguments to pass to pandoc, so we can dynamically add lua filters, which covers the motivation of #2347.

```r
output_format_dependency(
  name = "example",
  pre_processor = function(...) {
    pandoc_lua_filter_args(c("example1.lua", "example2.lua"))
  },
  post_processor = function(...) {},
)
```

Also, this PR should be helpful in the situations like #2314.
In #2314, the issue author suggests `knit` filed in the YAML front matter to modify the `render()` function.
This is a powerful solution, but limited to RStudio users...

My proposal is available for R users on any environments.

Note that there is a remaining work to implement `attach_output_format_dependencies()`, like `htmltools::attachDependencies()`, but this can be a future work.

### CODE REVIEW

**Comment by atusy** [comment]
An example Rmd

<details>

````
---
title: foo
output:
  html_document:
    keep_md: true
---

```{cat, engine.opts=list(file = "example.lua")}
print("====lua filter====")
```

```{r}
rmarkdown::output_format_dependency(
  name = "foo",
  pre_processor = function(...) {
    cat("====pre_processor====\n")
    pandoc_lua_filter_args("example.lua")
  },
  post_processor = function(metadata, input_file, output_file, ...) {
    cat("====post_processor====\n")
    output_file
  }
)
```
````
</details>

**Comment by cderv** [comment]
Wow this is an interesting idea - I need to look into that closer so it could be not right now as I am currently focus on Quarto 1.3 release. But as soon as I can make time for some rmarkdown work (which is completely due and should happen soon), I'll give this a try. Maybe @yihui will have more time before. 

Anyhow, generalized approach needs more thoughts than simple PR. Your contribution is really appreciated and I wanted to be honest with you with current situation so that you don't think this PR is ignored! Thank you very much for this @atusy !

**Comment by atusy** [comment]
No worries! I'm not hurry.

File: R/output_format.R [hunk 1]
```diff
@@ -815,3 +815,56 @@ citeproc_required <- function(yaml_front_matter,
...
+
+#' @export
+knit_print.output_format_dependency <- function(x, ...) {
+  knitr::asis_output(list(), meta = list(x))
+}
+
+merge_output_format_dependency <- function(fmt, dep) {
+  dep$name <- NULL
+  do.call(output_format,
+          c(dep, list(knitr = NULL, pandoc = list(), base_format = fmt)))
```
Comment (cderv) [inline R/output_format.R:858]:
```suggestion
          c(dep, list(knitr = list(), pandoc = list(), base_format = fmt)))
```
would this work too ?

**Review by cderv — COMMENTED** [review]
That is really clever way to generalize that ! Thank you.  

Should we completely make all the `output_format` argument available that way ? it seems something possible. 

At least, I would like to see if we can add a specific handling for some of the pandoc component. `args` and `lua_filters` are among the content one would want to modify 

I wonder if we shouldn't extend this also to the knitr and pandoc component currently set as `knitr = NULL` and `pandoc = list()`

This would mean being possible to do something like 
```r
output_format_dependency <- function(name,
                                     pre_processor = NULL,
                                     post_processor = NULL,
                                     file_scope = NULL,
                                     on_exit = NULL,
                                     knitr = list(),
                                     pandoc = list())
```

so that one can do 
```r
output_format_dependency(pandoc = list(lua_filters = "example.lua"))
```

Did you try making that works ? 

Obviously we need to add some unit test for this so that we are sure it merges as we expect, and we don't break it in the future. 

Otherwise, all good ! Really exciting work ! Thanks again !

**Comment by atusy** [comment]
@cderv

Thank you so much for your review! I am excited by your reactions!

I will try adding the pandoc argument and unit tests in this or next week.

> I wonder if we shouldn't extend this also to the knitr and pandoc component currently set as knitr = NULL and pandoc = list()

About the above comment,

- `knitr` is not supproted because
    - the timing of solving `output_format_dependency` is generally too late because `merge_output_format_dependencies()` is called **after knit()**.
    - users/developers already have full powers on controlling the options within chunks by setters and getters (e.g., `opts_knit$set()`)
    - Calling setters within `output_format_dependency()` can be a solution, but the timing of the effect can be ambious. Specifically, there is no way to controll chunk options of the **current** chunk https://github.com/yihui/knitr/issues/1798 .

- `pandoc` seems possible. Although I wrote it is not supported for some reason, maybe I was wrong here :smile: 

    > Supported arguments are pre_processor, post_processor, file_scope, on_exit.
    > The others are not because dependency can be solved after post_knit, which finalizes knit_meta.

**Comment by cderv** [comment]
> the timing of solving output_format_dependency is generally too late because merge_output_format_dependencies() is called after knit().

Oh indeed. I missed that. I was too exited by the new feature probably ! 😄 

> The others are not because dependency can be solved after post_knit, which finalizes knit_meta.

Interesting 🤔  - initial issue was to have a way to add Lua filter more easily as a dependency. I thought we could somehow add to the `pandoc = list(lua_filters = )` . Maybe this is the only element of Pandoc that we can use 

Otherwise, it is easy enough to do a wrapper of this new function, as lua_filters can be added  in pre_processor as  you did. But I did it another for some reason (I think to add a format lua filter last or something like that). 

I let you look into this and we'll see if we can change things to make it happen.

**Comment by atusy** [comment]
Okay, I confirmed `pandoc = list(lua_filters = ...)` works.
Here's an example Rmd that a filter echoes `====lua filter====` to stdout.

<details><summary>Rmd</summary>

````
---
title: foo
output:
  html_document:
    keep_md: true
---

```{cat, engine.opts=list(file = "example.lua")}
print("====lua filter====")
```

```{r}
rmarkdown::output_format_dependency(
  name = "foo",
  pandoc = list(
    lua_filters = "example.lua"
  ),
  pre_processor = function(...) {
    cat("====pre_processor====\n")
    # pandoc_lua_filter_args("example.lua")
  },
  post_processor = function(metadata, input_file, output_file, ...) {
    cat("====post_processor====\n")
    output_file
  }
)
```
````

</details>

<details><summary>stdout from render()</summary>
```
processing file: example.Rmd
                                                                                                                                                                                                                                                                                                                                                                                            
                            
output file: example.knit.md

====pre_processor====
/usr/bin/pandoc +RTS -K512m -RTS example.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output example.html --lua-filter /home/atusy/ghq/github.com/rstudio/rmarkdown/inst/rmarkdown/lua/pagebreak.lua --lua-filter /home/atusy/ghq/github.com/rstudio/rmarkdown/inst/rmarkdown/lua/latex-div.lua --lua-filter example.lua --embed-resources --standalone
 --variable bs3=TRUE --section-divs --template /home/atusy/ghq/github.com/rstudio/rmarkdown/inst/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmpxmaQr6/rmarkdown-strb7dc26a6dd7a3.html 
====lua filter====
====post_processor====

Output created: example.html
```

</details>

So the last task is to add tests.

**Comment by atusy** [comment]
To implement tests, I rebased my PR onto #2488 .
I will rebase again to main when #2488 is merged.

File: R/render.R [hunk 2]
```diff
@@ -840,6 +840,8 @@ render <- function(input,
 
     perf_timer_start("pre-processor")
 
+    output_format <- merge_output_format_dependencies(output_format, knit_meta)
+
```
Comment (cderv) [inline R/render.R:844]: I think we should make this conditional. This is safer than just relying on `merge_output_format_dependencies` to do nothing IMO. More clear also from `render.R` file 

```suggestion
    if (has_dependencies(knit_meta, "output_format_dependency")) {
        output_format <- merge_output_format_dependencies(output_format, knit_meta)
    }
```

What do you think ?

**Review by cderv — COMMENTED** [review]
All good. You can rebase on main and maybe make the suggested change, and all good to merge ! 

Thank you

File: R/render.R (hunk 2 above)
Comment (cderv) [reply R/render.R]: Yes I understand why you feel that. But thank you :)

**Review by cderv — APPROVED** [review]
Awesome ! 

Need to be AFK but I'll rebase later if you haven't done it and then I'll merge 

Thank you

**Comment by atusy** [comment]
rebased and added news.

**Comment by yihui** [comment]
@atusy Do you have time to write a short blog post on this new feature? One or two examples would be enough. I feel many people will be interested in this feature. Feel free to publish the post in your own blog (https://blog.atusy.net), and we can help promote it. Thanks!

**Comment by atusy** [comment]
Yeah sure. I would post one in the next week!

**Comment by atusy** [comment]
done https://blog.atusy.net/2023/08/28/rmarkdown-output-format-dependency/

**Pull request merged**